sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp

```diff
@@ -0,0 +1,122 @@
+//==------------------ matrix.hpp - SYCL matrix ----------------*- C++ -*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+
+#include <CL/__spirv/spirv_ops.hpp>
+#include <CL/sycl/detail/defines_elementary.hpp>
+#include <CL/sycl/feature_test.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental::matrix {
+
+enum class matrix_layout { row_major, col_major, packed_a, packed_b };
+
+template <matrix_layout Layout> struct spv_matrix_layout_traits {
+  static constexpr __spv::MatrixLayout value = __spv::MatrixLayout::RowMajor;
+};
+
+#define SPV_MATRIX_LAYOUT_TRAITS(LAYOUT, SPV_LAYOUT)                           \
+  template <> struct spv_matrix_layout_traits<LAYOUT> {                        \
+    static constexpr __spv::MatrixLayout value = SPV_LAYOUT;                   \
+  };
+
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::row_major,
+                         __spv::MatrixLayout::RowMajor)
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::col_major,
+                         __spv::MatrixLayout::ColumnMajor)
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::packed_a, __spv::MatrixLayout::PackedA)
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::packed_b, __spv::MatrixLayout::PackedB)
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          matrix_layout Layout = matrix_layout::row_major>
+struct joint_matrix {
+public:
+  __spv::__spirv_MatrixINTEL<T, NumRows, NumCols,
+                             spv_matrix_layout_traits<Layout>::value> *spvm;
+  joint_matrix(Group sg) {
+#ifndef __SYCL_DEVICE_ONLY__
+    (void)sg;
+    throw runtime_error("joint matrix is not supported on host device.",
+                        PI_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+  }
+};
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          matrix_layout Layout = matrix_layout::row_major,
+          access::address_space Space>
+inline __SYCL_ALWAYS_INLINE void joint_matrix_load(
+    Group sg, joint_matrix<Group, T, NumRows, NumCols, Layout> &res,
+    multi_ptr<T, Space> src, size_t stride, matrix_layout L = Layout) {
```

Comment:
Function argument `L` is not used. Probably we should use it as operand of `__spirv_MatrixLoadINTEL`, but it must be a constant.
It seems like we have two layouts here: one for the memory we read from, and another for the matrix object, right? If so, I think we need another template parameter to specify layout of the memory.

Reply:
__spirv_MatrixLoadINTEL  already takes layout as argument and this should be L, the memory layout.
Are we using the matrix layout for that argument? If yes this should be changed.

The matrix layout is kept on the matrix type itself.

Reply:
IMO `memL` and `matL` are not quite user friendly names. Are `memory_layout` and `matrix_layout` more readable?

Reply:
Right now, the user does not need to specify any template argument when calling load/store. If we move the memory layout to the template, the user will have to specify each of the parameters (deduce all or nothing) when calling load/store. 

The other reason is that I would like to not converge too much (unless there is a big reason) from Vulkan and WMMA APIs thus keeping the meory layout as argument to the functions

Reply:
I guess MemL is behind MatL in Template parameters list?